[export] check non-negative modulus, avoid unnecessary congruences, in export solver

pianpwk opened the pull request:
Fixes #144852 

Summary: if we find x = a * k + b, make sure a is non-negative. Also don't introduce unnecessary congruences on arbitrary floordiv expressions. Test case checks we don't introduce a divisiblity guard.

Test Plan: test_export

Differential Revision: D68245292




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

pianpwk commented:
> This is resolving #144852. With this fix (for the toy test case in this PR) we get:
> 
> ```
> torch._dynamo.exc.UserError: Constraints violated (size)! For more information, run with TORCH_LOGS="+dynamic".
>   - Not all values of size = L['args'][0][1].size()[0] in the specified range satisfy the generated guard 96 + (L['args'][0][1].size()[0] // 2) + (((-1)*(L['args'][0][1].size()[0] // 2)) // 2) <= 192.
>   - Not all values of size = L['args'][0][1].size()[0] in the specified range satisfy the generated guard 2 <= L['args'][0][1].size()[0] and L['args'][0][1].size()[0] <= 385
> Suggested fixes:
>   _size = Dim('_size', min=1, max=96)
>   size = 4*_size
> ```
> 
> However this is still recommending that the dimension is divisible by 4. Digging a bit deeper, it actually looks like this is introduced by the export solver here:
> 
> https://github.com/pytorch/pytorch/blob/926f9056a9b4a74b48e1ea7b43dd7e46177633df/torch/fx/experimental/symbolic_shapes.py#L2404-L2405
> 
> , and not an explicitly emitted guard.
> Going back to the toy example:
> 
> ```
>         class Foo(torch.nn.Module):
>             def forward(self, x, y):
>                 size = y.shape[0]
>                 h = size // 2
>                 start = (192 - h) // 2
>                 return x[start : start + h]
> 
>         inps = (torch.randn(192), torch.randn(48))
>         dynamic_shapes = {
>             "x": None,
>             "y": {0: Dim("size")},
>         }
> ```
> 
> Here, `start = (192 - s0 // 2) // 2`, `h = s0 // 2`, and the slice op emits the guard `start + h <= x.size[0]`, which resolves basically to `96 + s0//2 - s0//4 <= 192`. According to the docstring here,
> 
> https://github.com/pytorch/pytorch/blob/926f9056a9b4a74b48e1ea7b43dd7e46177633df/torch/fx/experimental/symbolic_shapes.py#L2389-L2392
> 
> , the floordiv handling that export does implicitly rewrites the `s0//4` term into `(s0 - 0) / 4`, and guards on `s0 % 4 == 0`.
> However I feel like this is unnecessary? The guard might be helpful with cases like floordivs in cases of reshapes (but there I feel like an explicit `Mod` guard is emitted?), but in this specific case the floordiv is just arbitrary int computation on the user side, that doesn't require any divisibility guarantees. I feel that the correct thing is to only guard on the `96 + s0//2 - s0//4` expression as whole, that the user has provided to us, and guarding on the `s0` subterm feels intrusive.
> 
> Removing the lines above where we add the congruence, the divisibility suggestion goes away:
> 
> ```
> torch._dynamo.exc.UserError: Constraints violated (size)! For more information, run with TORCH_LOGS="+dynamic".
>   - Not all values of size = L['args'][0][1].size()[0] in the specified range satisfy the generated guard 96 + (L['args'][0][1].size()[0] // 2) + (((-1)*(L['args'][0][1].size()[0] // 2)) // 2) <= 192.
>   - Not all values of size = L['args'][0][1].size()[0] in the specified range satisfy the generated guard 2 <= L['args'][0][1].size()[0] and L['args'][0][1].size()[0] <= 385
> Suggested fixes:
>   size = Dim('size', max=384)
> ```
> 
> cc @avikchaudhuri @angelayi

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2508,6 +2508,9 @@ def _reduce_congruences(self) -> Dict[sympy.Symbol, Set[sympy.Expr]]:
                     if isinstance(modulus, sympy.Integer) and isinstance(
                         remainder, sympy.Integer
                     ):
+                        # Make sure modulus is non-negative
```
I *think* this is OK (because it essentially flips `tmp` to another variable equal to `-tmp`?). But worth checking with divisor other than `2`. Mostly worried whether the remainder will get updated correctly. E.g., when divisor is `3`, `-1 = 2 mod 3`.

avikchaudhuri commented:
> The floordiv handling that export does implicitly rewrites the `s0//4` term into `(s0 - 0) / 4`, and guards on `s0 % 4 == 0`.
> However I feel like this is unnecessary? The guard might be helpful with cases like floordivs in cases of reshapes (but there I feel like an explicit `Mod` guard is emitted?), but in this specific case the floordiv is just arbitrary int computation on the user side, that doesn't require any divisibility guarantees. I feel that the correct thing is to only guard on the `96 + s0//2 - s0//4` expression as whole, that the user has provided to us, and guarding on the `s0` subterm feels intrusive.
> 
> Removing the lines above where we add the congruence, the divisibility suggestion goes away:

There's not much point in removing this line, because the floordiv and mod handlers are exactly how congruences are added to the solver, which feeds into the `reduce_congruences` algorithm. No congruences, no need for that part. So the more relevant question is, do we need congruence solving at all?

This made sense when we were still trying to simplify guards, because the Dim language was limited, there was no turning guards into runtime assertions, and we didn't want guards to get ignored. So we used handlers that looked at concrete values of divisors to simplify guards (make them stricter or force specialization).

With automatic dynamic shapes I think we can just move away from interpreting floordiv / mod completely. The online solver does what it does, whatever guards remain we just add as runtime assertions?

pianpwk committed:
[export] check non-negative modulus, avoid unnecessary congruences, in export solver (#144925)

Summary:

if we find x = a * k + b, make sure a is non-negative. Also don't guard on unnecessary congruences, especially if the floordiv is just from arbitrary symint compute. In the test case, check that we don't introduce divisibility on the dynamic dim.

Test Plan: test_export

Differential Revision: D68245292

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into export-D68245292

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into export-D68245292

pianpwk committed:
keep mod congruence

bobrenjc93 commented on test/export/test_export.py:
```diff
@@ -1252,6 +1252,37 @@ def false_fn(x):
+            )
+            size = new_shapes["y"][0]
+            self.assertTrue(isinstance(size, _Dim))
+            self.assertTrue(not isinstance(size, _DerivedDim))
```
@pianpwk I've been staring at this test for 5 minutes and it's still not obvious what's happening. Would you mind explaining and perhaps adding a comment to the code? At the very least maybe a reference to the issue?

Here's my train of thought in case it's helpful

1) the test is about negative mods, so I first looked for `%`, didn't find any and so now I'm guessing that it has something to do with the divisibility of either `h = size // 2` or `start = (192 - h) // 2`
2) the assert is very non obvious to me. at a high level you are try catching, refining the shapes from the error message and asserting shape[y][0] is Dim? But isn't it already a Dim? Or by default are dims DerivedDims?
3) Tracing through the code with example inputs...

x = [192, 1]
y = [48, 1]

size = u[0] = 48
h = size // 2 = 48 // 2 = 24
start = (192 - h) // 2 = (192 - 24) // 2 = 168 // 2 = 84
x[start : start + h] = x[84 : 84 + 24] = x[84 : 108]

where is the congruence divisibility symbolic reasoning happening?

pianpwk replied:
@bobrenjc93 Sorry, I realize this was probably terrible to try to understand.

With the non-negative check fix alone, the suggested fix would say something like:
```
_dy = Dim("size", max=...)
dy = 4*_dy
```
saying that this dim should be divisible by 4.

But there's nothing inherent in the program that requires this divisibility constraint at all, it's just the `floor_div_handler` function below adding a divisibility guard when we see any floordivs (in this case `(192 - size // 2) // 2`).

The test could probably be better written, but it just checks that we don't suggest a divisibility fix (by checking that a DerivedDim isn't being suggested).
